Read package-info.json and create archives

stertooy opened the pull request:
This should be a working and usable action now.

I've added an optional boolean input "dry-run" so people can run through all the steps except for the actual publishing on GitHub. If combined with the upload-artifact action, it's possible to manually check all of the assets that would be added to the GitHub release.

One major difference with the older script from ReleaseTools, is that this action will not fail if the release tag already exists, and will instead update the assets. This seems acceptable to me (maybe even useful in some cases?), but perhaps I'm not considering certain implications of this...?

stertooy commented on pkginfo_to_json.g:
```diff
@@ -7,9 +7,6 @@ function(o, x)
     PrintTo(o, "null");
 end);
 
-result := ValidatePackageInfo("PackageInfo.g");
```
I'm not sure I understand what you mean?

The json file is still added as an asset to the release (although I renamed it to `package-info.json`, to match the name used in `‎pkginfo_to_json.g`).

As for the quoted code: @fingolfin made a separate, earlier step for `ValidatePackageInfo` in 2853b41dc53e99a122f36df9765d3887f76f696f.

stertooy commented on action.yml:
```diff
@@ -20,41 +18,130 @@ runs:
+        files: ${{ env.ASSETS }}/*
+
+    - name: "Upload the release"
+      uses: actions/upload-artifact@v4
```
I'm not sure what you mean, version 4.6.2 seems to be the latest one? Unless you want to use a different action alltogether?

stertooy commented on action.yml:
```diff
@@ -20,41 +18,130 @@ runs:
+          echo "could not find release script ${{ inputs.release-script }}"
+          exit 1
+        fi
+        . "${{ inputs.release-script }}"
```
That might be better... but that part of the action is basically WIP, and mostly copied from the ReleaseTools script:
https://github.com/gap-system/ReleaseTools/blob/5237768ef60bdcb7ca62f43cb522a60ae356abd2/release-gap-package#L518-L532

I think I want a better understanding of what kind of "release scripts" we are dealing with, and whether it's possible/desirable to provide the functionality as part of this action instead, before I finish that part. Perhaps @fingolfin knows of some packages that use/need such script?

limakzi replied:
This info is very useful and I like your proposal of wider view of actions. I think it should be part of MVP too.

fingolfin commented:
Some more notes:

> this action will not fail if the release tag already exists, and will instead update the assets.

I agree.

> Running "autogen.sh" if a package contains it

This is certainly necessary. Of course we could just require everyone to do this as part of a "release script", but ideally besides running `./autogen.sh`, one should also make sure autoconf/automakers/libtool are installed; the directory `autom4te.cache` can be deleted afterwards; etc. -- so to me it seems nice to just do it for the user?

stertooy commented:
> This is still marked as a draft. What is missing before it is "ready for being considered for merging" ?

I still need to switch the package being used in the CI tests, but that won't work until that package has a release with a "package-info.json". That's all, I think.

limakzi commented on action.yml:
```diff
@@ -20,41 +19,131 @@ runs:
+        mv package-info.json $ASSETS
+        
+    - name: "Read data from package-info.json"
+      shell: bash
```
And adding these as outputs of this step, you mean?

fingolfin replied:
Sure, or like this step does right now (the script puts the "outputs" into env, via `$GITHUB_ENV`). Anyway, the current code works, so ... *shrug*

limakzi commented on action.yml:
```diff
@@ -20,41 +19,131 @@ runs:
+        esac
 
-    - name: Cleanup before making a tarball
+    - name: "Make copy of package"
```
I would not be afraid of this destructive action, to be fair. 👍🏻

fingolfin replied:
Well you added a comment to explain what happens. if it works, it works, and I am OK with leaving it now that I have a bit of an understanding why it's there. Thank you!

stertooy commented:
I added some functionality for release scripts and autogen, mostly copied from ReleaseTools again. Two small remarks:
- I didn't copy ReleaseTools' different handling when a file named ".release" is present. I assume it's fair to expect a user to put this script in the `release-script` input, and to ensure it deletes itself if so desired?
- Before running autogen, we install `autoconf automake build-essential cmake`. Not sure if this is enough, or too much, but it won't hurt either I guess.

All in all, should be ready for a "test drive" now. There's plenty of room for improvements, but I think this is about the right time to check if what we have now works well.

I also switched the CI to use ActionTestGAPDocPackage and ActionTestOldDocPackage instead of Example. However:
- Neither of these packages has an `autogen.sh` file, so that part of the action remains untested.
- Neither of these packages has a release script, so that part of the action remains untested.
- The `build-pkg-docs` action fails at building ActionTestOldDocPackage's documentation... Maybe I should comment out this package until that gets fixed?

fingolfin commented:
I just pushed a fix for ActionTestOldDocPackage

wilfwilson commented on action.yml:
```diff
@@ -1,17 +1,15 @@
+  dry-run:
+    description: "Set to true to not push the release to GitHub"
+    required: false
+    default: "false"
```
I agree with @fingolfin here btw.
